File: ztm_master_coding_interview_course.py/linked_lists/doubly_linked_list.py
from typing import Any

class Node:
    """ A class representing a node in a linked list. """
    def __init__(self, data):
        """ Initialize the node with the given data. """
        self.data = data
        self.previous = None
        self.next = None


class LinkedList:
    """ A class representing a doubly linked list. """
    def __init__(self):
        """ Initialize an empty doubly linked list. """
        self.head = None
        self.tail = None


    def prepend(self, data: Any) -> None:
        """ Add nodes to head. Time complexity O(1). """
        ADD_NODE = Node(data=data)
        if self.head is None:
            self.head = ADD_NODE
            self.tail = ADD_NODE
        else:
            ADD_NODE.next = self.head
            self.head.previous = ADD_NODE
            self.head = ADD_NODE


    def append_nodes(self, data: Any) -> None:
        """ Add a node with the given data to the end of the list. Time complexity O(1). """
        ADD_NODE = Node(data)

        if self.head is None:
            self.prepend(ADD_NODE)
        else:
            ADD_NODE.previous = self.tail
            self.tail.next = ADD_NODE
            self.tail = ADD_NODE


    def insert_node_at(self, index: int, data: Any) -> None:
        """ Insert a node with the given data at the specified index. Time complexity O(n). """
        ADD_NODE = Node(data=data)
        count = 0
        iterator = self.head

        if index < 0:
            raise IndexError("Positive integers only...")
        elif index == 0:
            self.prepend(data)

        while iterator is not None:
            if count == index:
                ADD_NODE.next = iterator
                ADD_NODE.previous = iterator.previous
                iterator.previous.next = ADD_NODE
                return

            count += 1
            iterator = iterator.next

        raise Exception("Sorry, linked list out of bounds...")


    def print_out_linked_list(self) -> str | bool:
        """
        Print out all nodes in the linked list.
        Last node references to a None value.
        Lastly, if Linked List is empty, return False value.
        """
        if self.head is None:
            print("Sorry, linked list does not has info :(")
            return False

        while self.head is not None:
            print(self.head.data, '<-->', end=' ')
            self.head = self.head.next
            if self.head is None:
                print(None, end=' ')

        print('\nExecution completed successfully :D')



if __name__ == '__main__':
    my_linked_list = LinkedList() # Instance of LinkedList class.

    # Add nodes to the beginning and end of the list.
    my_linked_list.prepend(10)
    my_linked_list.prepend(5)
    my_linked_list.prepend(15)
    my_linked_list.prepend(20)
    my_linked_list.append_nodes("Last")
    my_linked_list.append_nodes("bYE")

    # Insert a node at a specific index.
    my_linked_list.insert_node_at(5, "I am here")

    # Print out linked list.
    my_linked_list.print_out_linked_list()

File: ztm_master_coding_interview_course.py/linked_lists/doubly_linked_list.js
class Node {
    constructor (data) {
        this.data = data
        this.previous = null
        this.next = null
    }
}


class LinkedList {
    constructor () {
        this.head = null
        this.tail = this.head
        this.length = 0
    }


    prepend (data) {
        /* Add nodes to linked list head. */

        const ADD_NODE = new Node(data)

        if (this.head === null) {
            this.head = ADD_NODE
            this.tail = ADD_NODE
            this.length++
        }
        else {
            ADD_NODE.next = this.head
            this.head.previous = ADD_NODE
            this.head = ADD_NODE
            this.length++
        }
    }


    appendNodes (data) {
        /* Insert data at linked list's tail. Time complexity O(1). */

         const NEW_NODE = new Node(data)
        if (this.head === null) {
            this.prepend(data)
        }
        else {
            this.tail.next = NEW_NODE
            NEW_NODE.previous = this.tail
            this.tail = NEW_NODE
            this.length++
        }
    }


    insertNodeAt (index, data) {
        /* Insert new data at index given. Time complexity O(n). */

        const ADD_NODE = new Node(data) // Create new node.

        if (index < 0) {
            throw new Error('Invalid index...')
        } else if (index === 0) {
            this.prepend(data)
        }

        let count = 0
        let iterator = this.head

        while (iterator !== null) {
            if (count === index) {
                ADD_NODE.next = iterator // Set the new node point to current node.
                ADD_NODE.previous = iterator.previous // Set the current node's next pointer to the new node.
                iterator.previous.next = ADD_NODE
                this.length++
            return
            }
            count++
            iterator = iterator.next
        }
        throw new Error('Linked list out of bounds...')
    }


    removeNode (index) {
        /* Remove node at index given. Time complexity O(n) if the index is greater than zero, if it is equals to zero time complexity would be O(1). */

        if (index < 0) {
            throw new Error('Invalid index...')
        } else if (index === 0) {
            // If the index is 0, remove the head node and update the head reference.
            this.head = this.head.next
            return
        }

        let count = 0
        let iterator = this.head

        while (iterator !== null) {
            if (count === index - 1) {
                // If you want to checkout which item will be removed, uncomment the following line:
                //console.log("Node to remove ->", iterator.next.data);
                // Next node is the target node, remove it by updating the next reference of the previous node.
                iterator.next = iterator.next.next
                this.length--
            return
            }
            count++
            iterator = iterator.next
        }
        throw new Error('Linked list out of bounds...')
    }

    /*
    reverse () {
        // TODO...
        let counter = 0
        let auxiliary

        while (counter < this.length) {
            let currentNode = this.head
            let currentLastNode = this.tail
            currentLastNode = this.head
            this.head = this.tail
            currentNode = currentNode.next


            counter++
        }
    }
    */

    printOutLinkedList () {
        let iterator = this.head
        const llArray = [];

        while (iterator !== null) {
            llArray.push(iterator.data)
            iterator = iterator.next
        }
        return llArray
    }
}


// Example usage.
const MyLinkedList = new LinkedList()

MyLinkedList.prepend(5)
MyLinkedList.prepend(115)
MyLinkedList.prepend(1)
MyLinkedList.appendNodes(15)
MyLinkedList.appendNodes("Last")
MyLinkedList.insertNodeAt(2, "I am here")
MyLinkedList.removeNode(4)
// MyLinkedList.reverse()

console.log(MyLinkedList.printOutLinkedList())
